### DISCUSSION TITLE
MacOS  cloning as a way to dedupe files: how to detect already cloned files?

### DISCUSSION THREAD

**Discussion by lucbouge** [pkolaczk/fclones#276] [General]
I have discover the "cloning" facility of recent MacOS version: `cp -c`. Essentially, it copies a file by just sharing the data block on the disk, that is, without any space consumption. Only the metadata are specific, so that the two copies can live their own lives. 

If some block is modified at one of the files, then this very block is duplicated, and only it. It is called "Copy on write". It is of course ideal for files which are mostly read, as for instance archive files.

Is looks like the "fclones dedup" uses this feature, which is great.

However, a major drawback it that the "fclones dedup" does not realize that two copies of file **already** share the same blocks, so that they do not need to be deduplicated again, which make the repeated runs as long as the first one.

Would it be possible to optimize the "fclones dedup" strategy on MacOS to detect that two identical files are already "cloned", so that nothing is needed?

At least some years ago, it looked almost impossible: https://eclecticlight.co/2021/04/02/how-can-you-tell-whether-a-file-has-been-cloned-in-apfs/

Best regards,

Luc.

**Comment by kapitainsky** [comment]
And as you noted it is most likely close to impossible. APFS is close source file system with limited documentation. So unless Apple provides clear guidance/tool how to identify two full clones (as clones can be partial and files can be different) I think it is not even worth of time to try given limited gains it would provide.

**Comment by lucbouge** [comment]
Thanks, kapitainsky. I was expecting that things had moved since 2021. Too bad...

**Reply by lucbouge** [reply]
@kapitainsky 

I found this blog post which describes and implements a way to go. Did you know of it?
https://www.wolczko.com/nvm-blog/Clonefiles.pdf

**Reply by kapitainsky** [reply]
> @kapitainsky
> 
> I found this blog post which describes and implements a way to go. Did you know of it? https://www.wolczko.com/nvm-blog/Clonefiles.pdf

Thanks for sharing. Somebody did good job in investigating and documenting it. 

Now it needs somebody willing to implement it and submit PR.

Interesting question is if it is doable in rust as it uses some low level APIs.

**Comment by lucbouge** [comment]
Thanks. I will try to contact this guy and first test the software. Then, I will propose a PR. 

Actually, the software is just an executable file, easily compiled on Linux and MacOS. What is only needed from fclones is to test whether the command is reachable in the current execution environment, and if so, use it to compare potentially duplicated files.

Let me first contact the guy. I will keep you updated.

Regards,

Luc (From Brittany, France)

**Comment by lucbouge** [comment]
Dear Piotr @kapitainsky,

I contacted @pkolaczk and forked his extents project which detects clones files. With respect to the preliminary tests, it looks to work as expected.

Therefore, I would like to run larger tests, and fclones is an easy way to do it. Here is my plan.

When flcones detects that two files are likely to be similar (same length, same prefix, same suffix), it could offer a variety of means to finalize the detection. One of them is comparing byte-wise like diff, of course. Another would be to make a command/system call to an external program and catch the return code. One of those external programs could be `extents`, which compares the addresses and length of the physical blocks on the storing media and return 0 if all blocks are the same (same order, same address, same length).

I would like to propose you to fork your project and add this additional optional option to fclones, so that I can test it in the very large of my full backup disk (1 TB). Of course, you would be registered as a collaborator and the license would not be modified.

I would help me if you could insert at the right location the few lines of Rust to call an external program (`extents` with the two file descriptors and additional arguments), fetch its return code and interpret it as `CLONED` or `NOT_CLONED`. If the return code is `NOT_CLONED`, then it should go on with the regular fclones similarity detection code.

Regards, Luc.

**Reply by kapitainsky** [reply]
fclone is not my project:) I am only a small contributor.

Please note that this project is released under MIT license so you are pretty much free to do with it whatever you like.
